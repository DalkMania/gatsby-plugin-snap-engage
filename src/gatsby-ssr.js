import React from "react"
import { stripIndent } from "common-tags"

const getLanguage = (pathname, pluginOptions) => {
  const currentLocale = pathname.match(/^\/([^?\/]+)/)
  const locales = Object.keys(pluginOptions.locales)

  if((currentLocale !== null) && locales.filter(locale => locale == currentLocale)) {
    return pluginOptions.locales[currentLocale[1]]
  } else {
    return pluginOptions.locales[pluginOptions.defaultLocale]
  }
}

export const onRenderBody = ({ pathname, setPostBodyComponents }, pluginOptions) => {
  if (
    process.env.NODE_ENV === `production` ||
    pluginOptions.includeInDevelopment
  ) {
    const id = getLanguage(pathname, pluginOptions)

    return setPostBodyComponents([
      <script
        key={`gatsby-plugin-snap-engage`}
        dangerouslySetInnerHTML={{
          __html: stripIndent(`
          (function() {
            var se = document.createElement('script'); se.type = 'text/javascript'; se.async = true;
            se.src = '//storage.googleapis.com/code.snapengage.com/js/${id}.js';
            var done = false;
            se.onload = se.onreadystatechange = function() {
              if (!done&&(!this.readyState||this.readyState==='loaded'||this.readyState==='complete')) {
                done = true;
              }
            };
            var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(se, s);
          })();
          `),
        }}
      />,
    ])
  }
}