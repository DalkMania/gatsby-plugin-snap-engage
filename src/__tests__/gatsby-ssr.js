import { onRenderBody } from "../gatsby-ssr"

jest.mock(`common-tags`, () => {
    return {
      stripIndent: args => args,
    }
})

describe(`gatsby-plugin-snap-engage`, () => {
    describe(`onRenderBody`, () => {
        describe(`in development mode`, () => {
            it(`it does not add the script to the body, if the includeInDevelopment option is false`, () => {
                const pluginOptions = { includeInDevelopment: false }
                const setPostBodyComponents = jest.fn()
                const pathname = ''

                onRenderBody({ pathname, setPostBodyComponents }, pluginOptions)
                expect(setPostBodyComponents).not.toHaveBeenCalled()
            })

            it(`it does add the script to the body, if the includeInDevelopment option is true`, () => {
                const pluginOptions = { multilingual: false, id: '1', includeInDevelopment: true }
                const setPostBodyComponents = jest.fn()
                const pathname = ''

                onRenderBody({ pathname, setPostBodyComponents }, pluginOptions)
                expect(setPostBodyComponents.mock.calls).toMatchSnapshot()
                expect(setPostBodyComponents).toHaveBeenCalledTimes(1)
                expect(setPostBodyComponents).toHaveBeenCalledWith([
                    expect.objectContaining({ key: `gatsby-plugin-snap-engage` }),
                ])

                const result = JSON.stringify(setPostBodyComponents.mock.calls[0][0])
                expect(result).toContain(`code.snapengage.com/js/1.js`)
            })

        })

        describe(`in production mode`, () => {
            let env

            beforeAll(() => {
                env = process.env.NODE_ENV
                process.env.NODE_ENV = `production`
            })

            afterAll(() => {
                process.env.NODE_ENV = env
            })

            it(`it does add the script to the body`, () => {
                const pluginOptions = { multilingual: false, id: '1', includeInDevelopment: false }
                const setPostBodyComponents = jest.fn()
                const pathname = ''

                onRenderBody({ pathname, setPostBodyComponents }, pluginOptions)
                expect(setPostBodyComponents.mock.calls).toMatchSnapshot()
                expect(setPostBodyComponents).toHaveBeenCalledTimes(1)
                expect(setPostBodyComponents).toHaveBeenCalledWith([
                    expect.objectContaining({ key: `gatsby-plugin-snap-engage` }),
                ])

                const result = JSON.stringify(setPostBodyComponents.mock.calls[0][0])
                expect(result).toContain(`code.snapengage.com/js/1.js`)
            })

            it(`it does add the script to the body in multilingual mode, and uses the correct language code script from the locales array`, () => {
                const pluginOptions = { 
                    multilingual: true, 
                    defaultLocale: 'en',
                    locales: {
                    'en': 'english-script',
                    'fr': 'french-script'
                    }, 
                    includeInDevelopment: false 
                }
                const setPostBodyComponents = jest.fn()
                const pathname = '/fr/some-slug'

                onRenderBody({ pathname, setPostBodyComponents }, pluginOptions)
                expect(setPostBodyComponents.mock.calls).toMatchSnapshot()
                expect(setPostBodyComponents).toHaveBeenCalledTimes(1)
                expect(setPostBodyComponents).toHaveBeenCalledWith([
                    expect.objectContaining({ key: `gatsby-plugin-snap-engage` }),
                ])

                const result = JSON.stringify(setPostBodyComponents.mock.calls[0][0])
                expect(result).toContain(`code.snapengage.com/js/french-script.js`)
            })

            it(`it does add the script to the body in multilingual mode, and uses the correct language code script from the locales array based on base slug`, () => {
                const pluginOptions = { 
                    multilingual: true, 
                    defaultLocale: 'en',
                    locales: {
                    'en': 'english-script',
                    'fr': 'french-script'
                    }, 
                    includeInDevelopment: false 
                }
                const setPostBodyComponents = jest.fn()
                const pathname = '/fr'

                onRenderBody({ pathname, setPostBodyComponents }, pluginOptions)
                expect(setPostBodyComponents.mock.calls).toMatchSnapshot()
                expect(setPostBodyComponents).toHaveBeenCalledTimes(1)
                expect(setPostBodyComponents).toHaveBeenCalledWith([
                    expect.objectContaining({ key: `gatsby-plugin-snap-engage` }),
                ])

                const result = JSON.stringify(setPostBodyComponents.mock.calls[0][0])
                expect(result).toContain(`code.snapengage.com/js/french-script.js`)
            })

            it(`it does add the script to the body in multilingual mode, and uses the default language code script if the pathname doesn't match`, () => {
                const pluginOptions = { 
                    multilingual: true, 
                    defaultLocale: 'en',
                    locales: {
                    'en': 'english-script',
                    'fr': 'french-script'
                    }, 
                    includeInDevelopment: false 
                }
                const setPostBodyComponents = jest.fn()
                const pathname = '/some-slug'

                onRenderBody({ pathname, setPostBodyComponents }, pluginOptions)
                expect(setPostBodyComponents.mock.calls).toMatchSnapshot()
                expect(setPostBodyComponents).toHaveBeenCalledTimes(1)
                expect(setPostBodyComponents).toHaveBeenCalledWith([
                    expect.objectContaining({ key: `gatsby-plugin-snap-engage` }),
                ])

                const result = JSON.stringify(setPostBodyComponents.mock.calls[0][0])
                expect(result).toContain(`code.snapengage.com/js/english-script.js`)
            })
        })
    })
})