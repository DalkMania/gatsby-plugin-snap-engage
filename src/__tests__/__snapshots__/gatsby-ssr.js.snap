// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-plugin-snap-engage onRenderBody in development mode it does add the script to the body, if the includeInDevelopment option is true 1`] = `
Array [
  Array [
    Array [
      <script
        dangerouslySetInnerHTML={
          Object {
            "__html": "
          (function() {
            var se = document.createElement('script'); se.type = 'text/javascript'; se.async = true;
            se.src = '//storage.googleapis.com/code.snapengage.com/js/1.js';
            var done = false;
            se.onload = se.onreadystatechange = function() {
              if (!done&&(!this.readyState||this.readyState==='loaded'||this.readyState==='complete')) {
                done = true;
              }
            };
            var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(se, s);
          })();
          ",
          }
        }
      />,
    ],
  ],
]
`;

exports[`gatsby-plugin-snap-engage onRenderBody in production mode it does add the script to the body 1`] = `
Array [
  Array [
    Array [
      <script
        dangerouslySetInnerHTML={
          Object {
            "__html": "
          (function() {
            var se = document.createElement('script'); se.type = 'text/javascript'; se.async = true;
            se.src = '//storage.googleapis.com/code.snapengage.com/js/1.js';
            var done = false;
            se.onload = se.onreadystatechange = function() {
              if (!done&&(!this.readyState||this.readyState==='loaded'||this.readyState==='complete')) {
                done = true;
              }
            };
            var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(se, s);
          })();
          ",
          }
        }
      />,
    ],
  ],
]
`;

exports[`gatsby-plugin-snap-engage onRenderBody in production mode it does add the script to the body in multilingual mode, and uses the correct language code script from the locales array 1`] = `
Array [
  Array [
    Array [
      <script
        dangerouslySetInnerHTML={
          Object {
            "__html": "
          (function() {
            var se = document.createElement('script'); se.type = 'text/javascript'; se.async = true;
            se.src = '//storage.googleapis.com/code.snapengage.com/js/french-script.js';
            var done = false;
            se.onload = se.onreadystatechange = function() {
              if (!done&&(!this.readyState||this.readyState==='loaded'||this.readyState==='complete')) {
                done = true;
              }
            };
            var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(se, s);
          })();
          ",
          }
        }
      />,
    ],
  ],
]
`;

exports[`gatsby-plugin-snap-engage onRenderBody in production mode it does add the script to the body in multilingual mode, and uses the default language code script if the pathname doesn't match 1`] = `
Array [
  Array [
    Array [
      <script
        dangerouslySetInnerHTML={
          Object {
            "__html": "
          (function() {
            var se = document.createElement('script'); se.type = 'text/javascript'; se.async = true;
            se.src = '//storage.googleapis.com/code.snapengage.com/js/english-script.js';
            var done = false;
            se.onload = se.onreadystatechange = function() {
              if (!done&&(!this.readyState||this.readyState==='loaded'||this.readyState==='complete')) {
                done = true;
              }
            };
            var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(se, s);
          })();
          ",
          }
        }
      />,
    ],
  ],
]
`;
